text/2020-09-24-QoS.md

```diff
@@ -0,0 +1,172 @@
+# QoS: Quality of Service
+
+## Motivation
+
+Queries compete for resources and thus interfere with each other. Currently users can only deal with this in very time consuming ways by either increasing cluster capacity or altering their applications, the latter of which may take hours or days.
+
+Users want to ensure a quality of service for their queries. Some queries should be prioritized above others. For queries of the same priority, resources should be divided fairly among queries. When there are multiple tenants, provide resource isolation but still allow for high utilization.
+
+## Summary
+
+This solution provide QoS at the level of the TiKV node. QoS is configured both globally in PD and dynamically by clients.
+
+  * QoS Policy is set in PD for region groups such as key spaces (tenant) and tables
+    * Larger region groups have more capacity allocated
+  * Allow an application (TiDB) to create its own policies by sending a QoS-Request that further prioritizes its own capacity.
+  * Analytics queries can request a low QoS
+  * Apply local back pressure on a TiKV node by rejecting queries using too much capacity
+
+![QoS Architecture](../media/qos-architecture.png)
+
+![QoS Capacity Slicing](../media/media/QoS-capacity-slice.png)
+
+
+## Terminology
+
+* QoS: a relative priority setting. This is not a quota: usage is always “bursting” to achieve high utilization.
+* Capacity: the total resources available to be prioritized
+* Key Space: in a multi-tenant setup, every tenant gets a distinct key space. More generally a key space is designed for applications with different data ownership.
+
+
+## Detailed design
+
+### Architectural and Implementation advantages
+
+Ti Components are loosely coupled:
+  * PD stores policies and communicates them to TiKV
+  * TiKV performs query admission, providing localized back pressure
```

Comment:
I haven't thought about TiFlash. Managing QoS for the OLTP workload path is critical. For OLAP it is less important. TiFlash is also gaining some direct write support, but I have no idea how that works. As the MPP support for TiFlash improves it will be easier to handle TiFlash load by scaling out. Additionally, applications that benefit from TiFlash would generally be big enough to have their own cluster. This proposal will benefit smaller applications the most that *must* use a shared TiDB cluster until they grow larger.